### PR TITLE
feat(moshicam): add moshicam plugin

### PR DESCRIPTION
**Pull request by jarrel-b**
### Changes
This PR adds a plugin for https://moshi.cam/ The intended use is to promote mints of specific tokens from specific user contracts. 

Note: I wasn't sure if I need to add functions such as `getFees`, `simulateMint`, etc. I saw them in other plugins, but didn't see them as a requirement in the docs. I added `getMintIntent` since it was easier to implement.

### CODE REVIEW

**Review by Robinnnnn — APPROVED** [review]

**Comment by jarrel-b** [comment]
@Quazia @mmackz friendly ping we would like to starting boost ASAP!

**Comment by mmackz** [comment]
> @Quazia @mmackz friendly ping we would like to starting boost ASAP!

Will get back to you shortly after review. 

You will need to implement the `getProjectFees` and `simulateMint` functions in order for the project to be compatible with inbox. These functions are not documented yet, but I would be able to help implement them if needed.

Have you talked to anyone about this integration at all?

**Comment by jarrel-b** [comment]
Hey @mmackz sounds good will try to integrate them from examples. Haven't chatted with anyone on the team yet, lmk if it makes sense to set something up!

**Comment by jarrel-b** [comment]
@mmackz implemented `getProjectFees` and `simulateMint`. lmk what else is needed to get this over the line!

File: packages/moshicam/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "import": "./dist/esm/index.js",
+    "types": "./dist/types/index.d.ts"
+  },
+  "main": "./dist/cjs/index.js",
+  "module": "./dist/esm/index.js",
+  "description": "Plugin for moshicam",
+  "scripts": {
+    "bench": "vitest bench",
+    "bench:ci": "CI=true vitest bench",
+    "build": "pnpm build:types && pnpm build:esm && pnpm postbuild:esm && pnpm build:cjs && pnpm postbuild:cjs && BABEL_ENV=esmBundled pnpm rollup -c",
```
Comment (mmackz) [inline packages/moshicam/package.json:16]: We are using vite now for the build process so you will need to update the package.json and create a vite config file.

You can copy these files from another package and update the name.

You can see what needs to be changed here: https://github.com/rabbitholegg/questdk-plugins/pull/429/files

File: packages/moshicam/src/test-transactions.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+export const COLLECT_FROM_USER_MOSHICAM: TestParams<MintActionParams> = {
+  transaction: {
+    chainId: Chains.BASE,
+    from: '0xA064DB8fbFe520664442ee0E2650cc1C6c8f971e',
+    hash: '0x619762bdbded2725b58d62a164e52aa7a340e5ca64e37a6b8aeea4b63e9ff877',
+    input:
+      '0xacc10f11000000000000000000000000c46dd4c165d174ab5ca96026f3fa59855e73bcca00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001',
+    to: '0x704b204B48011b4B4554bf5659eFfeD1414CDa83',
+    value: '0.0001 ether',
```
Comment (mmackz) [inline packages/moshicam/src/test-transactions.ts:19]: Use wei here please.

100000000000000

File: packages/moshicam/src/Moshicam.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  encodeFunctionData,
+  type SimulateContractReturnType,
+  createPublicClient,
+  http,
+  parseEther,
+  type PublicClient,
+} from 'viem'
+import {
+  DEFAULT_MINT_PRICE,
+} from "./constants"
```
Comment (mmackz) [inline packages/moshicam/src/Moshicam.ts:18]: Run the formatter command `pnpm run format`

File: packages/moshicam/src/abi.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    stateMutability: 'payable',
+  },
+  {
+    type: 'function',
+    inputs: [],
+    name: 'mintPrice',
+    outputs: [{ name: '', internalType: 'uint256', type: 'uint256' }],
+    stateMutability: 'view',
+  },
+] as const
```
Comment (mmackz) [inline packages/moshicam/src/abi.ts:20]: Could you remove this `as const` assertion. Its causing a type error on the simulate contract return. https://github.com/gallery-so/questdk-plugins/blob/1ec752a0d360d4ba65b4b428dec4c04d5bfadf88/packages/moshicam/src/Moshicam.ts#L132

File: packages/moshicam/src/Moshicam.test.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    const result = await simulateMint(
+      mint,
+      DEFAULT_MINT_PRICE,
+      MOSHIMINTER_ADMIN,
+    )
+
+    expect(result.request.account.address).toEqual(MOSHIMINTER_ADMIN)
+    expect(result.request.address).toEqual(
```
Comment (mmackz) [inline packages/moshicam/src/Moshicam.test.ts:163]:
```suggestion
    expect(result.request.account?.address).toEqual(MOSHIMINTER_ADMIN)
```

File: packages/moshicam/src/Moshicam.test.ts [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+          const filter = await mint(params)
+          expect(apply(transaction, filter)).to.be.false
+        })
+      })
+    })
+  })
+})
+
+describe('Given the getMintIntent function', () => {
+  const { transaction } = COLLECT_FROM_USER_MOSHICAM
```
Comment (mmackz) [inline packages/moshicam/src/Moshicam.test.ts:60]: 'transaction' is assigned a value but never used.

File: packages/moshicam/src/abi.ts (hunk 4 above)
Comment (jarrel-b) [reply packages/moshicam/src/abi.ts:20]: Ahh ty this was what was causing the build error, I was scratching my head figuring this one out.

**Comment by jarrel-b** [comment]
Thanks for the review @mmackz, addressed the changes. Gonna investigate the build failures in a bit!

**Comment by mmackz** [comment]
> Thanks for the review @mmackz, addressed the changes. Gonna investigate the build failures in a bit!

Add this to your packages `package.json`

```
"types": "./dist/types/index.d.ts",
"typings": "./dist/types/index.d.ts",
 ```

**Comment by mmackz** [comment]
Plugin code is looking pretty good to me, once we get the build issues sorted it should be ready to merge.

Do you have a contact on telegram I can send to our BD team?

You can send it here or message me on telegram `@mattie_eth`

**Comment by jarrel-b** [comment]
> Plugin code is looking pretty good to me, once we get the build issues sorted it should be ready to merge.
> 
> Do you have a contact on telegram I can send to our BD team?
> 
> You can send it here or message me on telegram `@mattie_eth`

Excellent dm'ed you on telegram. Also fixed the build 🚀

**Comment by mmackz** [comment]
Still running into a few errors with the GH action.

Could you downgrade your PNPM to version 8.15.8 and regenerate the lockfile?

I tried pushing the lockfile to your branch, but dont have write access.

**Comment by jarrel-b** [comment]
Thanks @mmackz, I regenerated the lock file with. Mind kicking off the action again?

**Comment by jarrel-b** [comment]
Noticed the running `pnpm run test:ci -- --shard=1/3` locally fails with:
```
@rabbitholegg/questdk-plugin-okutrade:test:ci:  ❯ src/OkuTrade.test.ts  (0 test)
@rabbitholegg/questdk-plugin-polygon:test:ci:
@rabbitholegg/questdk-plugin-polygon:test:ci: > @rabbitholegg/questdk-plugin-polygon@1.0.0-alpha.55 test:ci /Users/jarrelbiscocho/dev/questdk-plugins/packages/polygon
@rabbitholegg/questdk-plugin-polygon:test:ci: > CI=true vitest --coverage "--shard=1/3"
@rabbitholegg/questdk-plugin-polygon:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci: ⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯
@rabbitholegg/questdk-plugin-okutrade:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci:  FAIL  src/OkuTrade.test.ts [ src/OkuTrade.test.ts ]
@rabbitholegg/questdk-plugin-okutrade:test:ci: TypeError: Cannot read properties of undefined (reading 'contracts')
@rabbitholegg/questdk-plugin-okutrade:test:ci:  ❯ Object.<anonymous> ../../node_modules/.pnpm/@gfxlabs+oku-chains@1.1.61_viem@2.13.1/node_modules/@gfxlabs/oku-chains/dist/index.js:1897:23
@rabbitholegg/questdk-plugin-okutrade:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
@rabbitholegg/questdk-plugin-okutrade:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci:  Test Files  1 failed (1)
@rabbitholegg/questdk-plugin-okutrade:test:ci:       Tests  no tests
@rabbitholegg/questdk-plugin-okutrade:test:ci:    Start at  11:18:01
@rabbitholegg/questdk-plugin-okutrade:test:ci:    Duration  1.71s (transform 650ms, setup 1ms, collect 0ms, tests 0ms, environment 23ms, prepare 144ms)
@rabbitholegg/questdk-plugin-okutrade:test:ci:
@rabbitholegg/questdk-plugin-okutrade:test:ci:  % Coverage report from v8
@rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
@rabbitholegg/questdk-plugin-okutrade:test:ci: File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
@rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
@rabbitholegg/questdk-plugin-okutrade:test:ci: All files          |   71.52 |      100 |       0 |   71.52 |
@rabbitholegg/questdk-plugin-okutrade:test:ci:  OkuTrade.ts       |   29.92 |      100 |       0 |   29.92 | ...11-115,118-127
@rabbitholegg/questdk-plugin-okutrade:test:ci:  constants.ts      |     100 |      100 |     100 |     100 |
@rabbitholegg/questdk-plugin-okutrade:test:ci:  ...n-addresses.ts |     100 |      100 |     100 |     100 |
@rabbitholegg/questdk-plugin-okutrade:test:ci:  utils.ts          |   42.75 |      100 |       0 |   42.75 | ...25-130,133-138
@rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
@rabbitholegg/questdk-plugin-okutrade:test:ci:  ELIFECYCLE  Command failed with exit code 1.
```
Will see if I can fix!

**Comment by mmackz** [comment]
> Noticed the running `pnpm run test:ci -- --shard=1/3` locally fails with:
> 
> ```
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  ❯ src/OkuTrade.test.ts  (0 test)
> @rabbitholegg/questdk-plugin-polygon:test:ci:
> @rabbitholegg/questdk-plugin-polygon:test:ci: > @rabbitholegg/questdk-plugin-polygon@1.0.0-alpha.55 test:ci /Users/jarrelbiscocho/dev/questdk-plugins/packages/polygon
> @rabbitholegg/questdk-plugin-polygon:test:ci: > CI=true vitest --coverage "--shard=1/3"
> @rabbitholegg/questdk-plugin-polygon:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci: ⎯⎯⎯⎯⎯⎯ Failed Suites 1 ⎯⎯⎯⎯⎯⎯⎯
> @rabbitholegg/questdk-plugin-okutrade:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  FAIL  src/OkuTrade.test.ts [ src/OkuTrade.test.ts ]
> @rabbitholegg/questdk-plugin-okutrade:test:ci: TypeError: Cannot read properties of undefined (reading 'contracts')
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  ❯ Object.<anonymous> ../../node_modules/.pnpm/@gfxlabs+oku-chains@1.1.61_viem@2.13.1/node_modules/@gfxlabs/oku-chains/dist/index.js:1897:23
> @rabbitholegg/questdk-plugin-okutrade:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
> @rabbitholegg/questdk-plugin-okutrade:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  Test Files  1 failed (1)
> @rabbitholegg/questdk-plugin-okutrade:test:ci:       Tests  no tests
> @rabbitholegg/questdk-plugin-okutrade:test:ci:    Start at  11:18:01
> @rabbitholegg/questdk-plugin-okutrade:test:ci:    Duration  1.71s (transform 650ms, setup 1ms, collect 0ms, tests 0ms, environment 23ms, prepare 144ms)
> @rabbitholegg/questdk-plugin-okutrade:test:ci:
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  % Coverage report from v8
> @rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
> @rabbitholegg/questdk-plugin-okutrade:test:ci: File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
> @rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
> @rabbitholegg/questdk-plugin-okutrade:test:ci: All files          |   71.52 |      100 |       0 |   71.52 |
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  OkuTrade.ts       |   29.92 |      100 |       0 |   29.92 | ...11-115,118-127
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  constants.ts      |     100 |      100 |     100 |     100 |
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  ...n-addresses.ts |     100 |      100 |     100 |     100 |
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  utils.ts          |   42.75 |      100 |       0 |   42.75 | ...25-130,133-138
> @rabbitholegg/questdk-plugin-okutrade:test:ci: -------------------|---------|----------|---------|---------|-------------------
> @rabbitholegg/questdk-plugin-okutrade:test:ci:  ELIFECYCLE  Command failed with exit code 1.
> ```
> 
> Will see if I can fix!

Its a issue with a different package. Ill get it sorted out.

**Comment by mmackz** [comment]
@jarrel-b If you pull in the latest changes from main, it should hopefully pass 🤞

**Comment by jarrel-b** [comment]
> @jarrel-b If you pull in the latest changes from main, it should hopefully pass 🤞

Awesome! Verified that the build passed for me locally.

**Review by mmackz — APPROVED** [review]

**Comment by jarrel-b** [comment]
lfg thanks @mmackz for helping get this merged!

**Pull request merged**